File: examples/go/ffs.go
package ffs

func ffs_ref(word uint32) uint32 {
	var zero uint32 = 0
	if word == zero {
		return 0
	}
	var cnt uint32
	var n32 uint32 = 32
	var one uint32 = 1
	for cnt = 0; cnt < n32; cnt++ {
		if ((one << cnt) & word) != zero {
			return cnt + one
		}
	}
	return 0
}

func ffs_imp(i uint32) uint32 {
	var n uint32 = 1
	var zero uint32 = 0
	var cffff uint32 = 0xffff
	var c00ff uint32 = 0x00ff
	var c000f uint32 = 0x000f
	var c0003 uint32 = 0x0003
	var n16 uint32 = 16
	var n8 uint32 = 8
	var n4 uint32 = 4
	var n2 uint32 = 2
	var n1 uint32 = 1

	if (i & cffff) == zero {
		n = n + n16
		i = i >> n16
	}
	if (i & c00ff) == zero {
		n = n + n8
		i = i >> n8
	}
	if (i & c000f) == zero {
		n = n + n4
		i = i >> n4
	}
	if (i & c0003) == zero {
		n = n + n2
		i = i >> n2
	}
	if i != zero {
		return (n + ((i + n1) & n1))
	} else {
		return 0
	}
}

File: doc/tutorial/code/ffs.rs
pub fn ffs_ref(w : u32) -> u32 {
    let mut i = 0;
    let mut word = w;
    if word == 0 { return 0; }

    let mut cnt = 0;
    while cnt < 32 {
        if (1 << i) & word != 0 {
            i += 1;
            return i;
        }
        i += 1;
        cnt += 1;
    }
    0
}

pub fn ffs_imp(j : u32) -> u32 {
    let mut n : u8 = 1;
    let mut i = j;
    if (i & 0xffff) == 0 {
        n += 16; i >>= 16;
    }
    if (i & 0x00ff) == 0 {
        n += 8; i >>= 8;
    }
    if (i & 0x000f) == 0 {
        n += 4; i >>= 4;
    }
    if (i & 0x0003) == 0 {
        n += 2; i >>= 2;
    }
    if i != 0 {
        return (n as u32) + ((i + 1) & 0x01);
    }
    else {
        return 0;
    }
}

fn main() {}
